### PR TITLE
HPCC-21142 Add persistence options to Java embed

### PR DESCRIPTION
**Pull request by richardkchapman**
HPCC-21142 Add persistence options to Java embed

Also fixes:

HPCC-21052 Use thread local variables for IEmbedFunctionContext
HPCC-21137 Problems with embedded java tests on master
HPCC-21127 Incorrect name mangling rules (additional fix)

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [x] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Mostly manual

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review

File: plugins/javaembed/HpccClassLoader.java [hunk 1]
```diff
@@ -32,7 +34,13 @@ private HpccClassLoader(java.net.URL [] urls, ClassLoader parent, int _bytecodeL
     }
     public Class<?> findClass(String className) throws ClassNotFoundException
     {
-        return defineClassForEmbed(bytecodeLen, bytecode, className.replace(".","/"));
+        Class result = (Class) classes.get(className);
```
Comment (ghalliday) [inline plugins/javaembed/HpccClassLoader.java]: does this need to be thead safe?

File: testing/regress/ecl/javaembed.ecl [hunk 2]
```diff
@@ -193,17 +193,18 @@ public class myClass {
...
   };
-  public static MyRecord transform(MyRecord in, int lim)
+  public static MyRecord testTransform(MyRecord in, int lim)
   {
     return new MyRecord(in.bfield, lim, in.dfield);
   }
 }
 ENDEMBED;
-
+/*
```
Comment (ghalliday) [inline testing/regress/ecl/javaembed.ecl]: should remove the comments

**Comment by ghalliday** [comment]
Also spotted that writeSimpleType, writeEnumType needs to call ReleaseStringUTFChars

**Review by ghalliday — COMMENTED** [review]
@richardkchapman a few comments

File: ecl/hqlcpp/hqlcpp.cpp [hunk 3]
```diff
@@ -11997,12 +11997,44 @@ void HqlCppTranslator::buildScriptFunctionDefinition(BuildCtx &ctx, IHqlExpressi
     funcctx.associateExpr(codeContextMarkerExpr, codeContextMarkerExpr);
     funcctx.associateExpr(globalContextMarkerExpr, globalContextMarkerExpr);
 
+    IHqlExpression *optionsParam = nullptr;
+    IHqlExpression *queryParam = nullptr;
+    unsigned numRealParams = 0;
+    ForEachChild(formalIdx, formals)
+    {
+        IHqlExpression *formal = formals->queryChild(formalIdx);
+        if (formal->queryId()==__optionsId)
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply ecl/hqlcpp/hqlcpp.cpp]: Case-sensitive is fine here. Looking for a special name added by codegen earlier.

File: plugins/javaembed/javaembed.cpp [hunk 4]
```diff
@@ -3458,9 +3339,263 @@ class JavaEmbedImportContext : public CInterfaceOf<IEmbedFunctionContext>
...
+            switch (persistMode)
+            {
+            case persistGlobal:
+                scopeKey.set("global");
+                break;
+            case persistWorkunit:
+                engine->getQueryId(scopeKey, true);
+                break;
+            case persistQuery:
+                engine->getQueryId(scopeKey, true);
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply plugins/javaembed/javaembed.cpp]: It should say false. I have changed.

File: plugins/javaembed/HpccClassLoader.java (hunk 1 above)
Comment (richardkchapman) [reply plugins/javaembed/HpccClassLoader.java]: I wondered the same thing, and concluded that it PROBABLY doesn't need to be at the moment as a classloader object is not shared by more than one function context. But that may change in the future and probably smart to put a synchronized on the method now.

File: testing/regress/ecl/javaembed.ecl (hunk 2 above)
Comment (richardkchapman) [reply testing/regress/ecl/javaembed.ecl]: Ok, though that will show up some more issues...

**Pull request merged**